src/codegen/dispatch.rs

```diff
@@ -358,3 +381,116 @@ pub(super) fn constructor_dispatch(
 
     cfg
 }
+
+/// On Solana, prepare the data account after deploy; ensure the account is
+/// large enough and write magic to it to show the account has been deployed.
+fn solana_deploy(
+    contract_no: usize,
+    vartab: &mut Vartable,
+    cfg: &mut ControlFlowGraph,
+    ns: &mut Namespace,
+) {
+    // Make sure that the data account is large enough. Read the size of the
+    // account via `tx.accounts[0].data.length`.
+    let account_length = Expression::Builtin(
+        Loc::Codegen,
+        vec![Type::Uint(32)],
+        Builtin::ArrayLength,
+        vec![Expression::StructMember(
+            Loc::Codegen,
+            Type::DynamicBytes,
+            Box::new(Expression::Subscript(
+                Loc::Codegen,
+                Type::Struct(StructType::AccountInfo),
+                Type::Array(
+                    Box::new(Type::Struct(StructType::AccountInfo)),
+                    vec![ArrayLength::Dynamic],
+                ),
+                Box::new(Expression::Builtin(
+                    Loc::Codegen,
+                    vec![Type::Array(
+                        Box::new(Type::Struct(StructType::AccountInfo)),
+                        vec![ArrayLength::Dynamic],
+                    )],
+                    Builtin::Accounts,
+                    vec![],
+                )),
+                Box::new(Expression::NumberLiteral(
+                    Loc::Codegen,
+                    Type::Uint(32),
+                    BigInt::zero(),
+                )),
+            )),
+            2,
+        )],
+    );
```

Comment:
I wonder if someday we can simplify this expression. It is too complex to be understood easily.

Reply:
> I wonder if someday we can simplify this expression. It is too complex to be understood easily.

As a start all the `Box::new()` could be replaced by `.into()`, I think this would help it already

Reply:
> As a start all the `Box::new()` could be replaced by `.into()`, I think this would help it already

Very nice, I did not realise that. Thanks @xermicus

Reply:
> I wonder if someday we can simplify this expression. It is too complex to be understood easily.

I have been wondering about this. We are generating a lot more CFG, and we will be generating more in the future. I am sure some rust macros would help here, but I'm ensure how to approach this.

We could have a rust `quote! {}` style approach, so this could be written:
```rust
let expr = sol_expr! { tx.accounts[0].data.length };
```
However implementing this is hard: this would mean:
 - Exporting the expression parser (possible)
 - Using sema on the exprression
 - Using cfg expression on the result

This could even work on statements.

Alternatively we could have some simple `macro_rules! {}` which would aid in the build-up of cfg expressions/instructions. I am not sure what they would look like, suggestions welcome.

Reply:
Macro rules would be an easier start, maybe something like I did a short while ago in the emit?

It would be awesome if we would have a proc macro for generating the whole tree, but I fear it is a lot of work we have no capacity for just yet...